sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -146,15 +146,15 @@ AccessToken AzureCliCredential::GetToken(
   auto const scopes = TokenCredentialImpl::FormatScopes(tokenRequestContext.Scopes, false, false);
   auto const tenantId
       = TenantIdResolver::Resolve(m_tenantId, tokenRequestContext, m_additionallyAllowedTenants);
+  auto const command = GetAzCommand(scopes, tenantId);
```

Comment:
In general, it's better to define variables as close to their consumption as possible.

In this case, I'm not sure I see why the `command` variable is declared outside the lambda for GetToken.

Reply:
I agree with the principle you mentioned but there are a few things going on here which makes this move worth it.

In this case, I am balancing that with bringing validation as close to the relevant variables as possible (which was the primary concern). I was initially going to validate right after the call to `Resolve` but keeping it in `GetAzCommand` but moving that function call closer achieved the best middle ground.

It also means only the command string is captured in the lambda (the relevant component), rather than two strings (scopes and tenantId), neither of which are needed by it other than generating the command.

There is another significant benefit in doing so, in that the invalid scopes/tenant ids never end up in the token cache and pollute it unnecessarily, whereas keeping it in the `GetToken` lambda pushes it into the token cache (requiring more frequent clean-up), even though `GetAzCommand` will eventually throw.

Reply:
But I don't see any validation done on command. I see validation on tenantId, but that is done before this.

What am I missing?

Reply:
Yes, the validation on tenantId happens in the implementation of `GetAzCommand`. We want to do this validation and throw **before** calling the lambda with the command string, to avoid putting the invalid scope/tenantid within the token cache.